Allocate bytebuf without magazine lock when threads get collisions

laosijikaichele opened the pull request:
Motivation:

For the `AdaptivePoolingAllocator`, thread picks a magazine based on the thread id. 

It's possible that multi-threads pick the same magazine, typically when total threads counts on `AdaptivePoolingAllocator` exceed `AdaptivePoolingAllocator.MAX_STRIPES`, this will likely cause threads collisions on acquiring the same magazine lock, which will hurt performance. 

To improve performance, we can try to allocate `bytebuf` without magazine lock when threads get collision.

Modification:

When threads get collision on allocating `bytebuf`, try to allocate from the `NEXT_IN_LINE` or `centralQueue`, which does not need holding the magazine's lock.

Result:

On my computer, the `AdaptivePoolingAllocator.MAX_STRIPES` value is 16, so I set threads number to 32 on the JMH methods with: `@Threads(32)`, which will make multi-threads collision on the same magazine easily happen.

JMH benchmark class: [AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark](https://github.com/netty/netty/blob/af1c11065fd34f865154ae59ef1771a4741c1bac/microbench/src/main/java/io/netty/microbench/buffer/AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.java)

Before PR:
```
Benchmark                                                                         (size)   Mode  Cnt         Score         Error  Units
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00064  thrpt   20  46739546.208 ± 2160446.387  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00256  thrpt   20  47588931.166 ± 1381959.617  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   01024  thrpt   20  45249206.622 ± 1402279.203  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   04096  thrpt   20  42326001.800 ±  850575.489  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00064  thrpt   20  41291766.588 ± 2900577.209  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00256  thrpt   20  44765730.938 ± 1015982.545  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     01024  thrpt   20  47533832.397 ± 4210977.985  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     04096  thrpt   20  43889158.191 ± 2304917.563  ops/s

[358.474s][info   ][gc,heap,exit] Heap
[358.474s][info   ][gc,heap,exit]  garbage-first heap   total 266240K, used 51436K [0x0000000700000000, 0x0000000800000000)
[358.474s][info   ][gc,heap,exit]   region size 2048K, 25 young (51200K), 3 survivors (6144K)
[358.474s][info   ][gc,heap,exit]  Metaspace       used 12310K, committed 12544K, reserved 1114112K
[358.474s][info   ][gc,heap,exit]   class space    used 1506K, committed 1600K, reserved 1048576K
```
After PR:
```
Benchmark                                                                         (size)   Mode  Cnt         Score         Error  Units
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00064  thrpt   20  57809451.614 ± 2048986.729  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00256  thrpt   20  56674394.370 ± 1417278.749  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   01024  thrpt   20  49802027.805 ± 1380102.638  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   04096  thrpt   20  51298374.215 ±  890586.221  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00064  thrpt   20  48191848.295 ± 1178419.524  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00256  thrpt   20  48231634.875 ± 4475455.680  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     01024  thrpt   20  52460274.226 ± 1018550.798  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     04096  thrpt   20  49182390.329 ± 1455722.168  ops/s

[360.090s][info   ][gc,heap,exit] Heap
[360.090s][info   ][gc,heap,exit]  garbage-first heap   total 266240K, used 50938K [0x0000000700000000, 0x0000000800000000)
[360.090s][info   ][gc,heap,exit]   region size 2048K, 24 young (49152K), 3 survivors (6144K)
[360.090s][info   ][gc,heap,exit]  Metaspace       used 12320K, committed 12544K, reserved 1114112K
[360.090s][info   ][gc,heap,exit]   class space    used 1506K, committed 1600K, reserved 1048576K
```
We can see relatively 7%~24% performance improvement.

This approach will not record the size histograms as it does not hold the magazine lock.

It's reasonable to not record the size on this scenario:
1. Because compute the preferred chunk size is a somewhat expensive operation, so we skip it when threads get collisions, which means at the moment, it's more important to reduce the contention overhead than record the size histograms.
2. If we successfully allocated `bytebuf` without lock, which means the size is not too much deviated than current histograms, which make it less urgent to re-calculate the size.
3. If we failed allocated `bytebuf` without lock, then it will continue to try to acquire the magazine lock and will record the size once it get the lock.

normanmaurer commented:
@laosijikaichele nice work!

franz1981 commented:
I'm trying to understand the drawbacks here, so let me ask some questions...
1. if the lock is not acquired and NEXT in line is available to be used, it means that a whole `Chunk` of some capacity is used, which is good
2. what happen to the acquired chunk (let's say 1 MB chunk - for 1K allocation)? 

My understanding will say: it will fully return once is released - but this assume that release happen immediately which is not always the case...
If the same thread has to further perform other allocations (let's say it has performed a write and the network is congested...i.e. the chunk cannot be released back) - it cannot make use of the same chunk and will compete again for a new one (or the one which it didn't acquired before!).
This means as well that such chunk is not available to the colliding other threads which probably could reuse it and amortize the cost of owning it - thanks to the size hint collected (assuming to be accurate to predict the allocation rate/size).
Makes sense?

chrisvest commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -569,7 +569,40 @@ public boolean tryAllocate(int size, int sizeBucket, int maxCapacity, AdaptiveBy
+                curr = null;
+            } finally {
+                if (curr != null) {
+                    // Something went wrong, let's ensure we not leak the newChunk.
```
```suggestion
                    // Something went wrong, let's ensure we not leak the 'curr' chunk.
```

chrisvest reviewed (commented):
This reduces the likelihood of expanding the magazine array. If we are generally able to maintain on-par or better performance with fewer magazines, then overall that's a win, as we in theory will do the same work with less memory.

I'm less convinced about pulling from the central queue. Those chunks are less likely to be used "soon" relative to any magazine buffers, and the central queue is, I think, very likely to have chunks available. The central queue is also a place where we'd consider trimming off excess memory and return it to the system, but we can't do that if we've nibbled some allocations off those chunks.

Have you tried running the benchmark with the 'central queue' bit removed?

laosijikaichele commented:
Thanks for the review, I will do more tests to answer the questions rose by @chrisvest and @franz1981. 

Besides, there are some points I want to discuss:
1. Based on my benchmark on `AdaptivePoolingAllocator`, the main bottleneck of the performance is the threads collisions on acquiring the magazine lock, back to the JMH benchmark case, if we change the `AdaptivePoolingAllocator.MAX_STRIPES` value to 32, which equals with the JMH benchmark threads number, it will make a far better performance results than current PR: relatively 100% performance improvements. So if we enlarge the threshold value of `AdaptivePoolingAllocator.MAX_STRIPES` to reduce threads collisions, then the performance will be better, but this comes a question, how to choose a proper MAX value of `AdaptivePoolingAllocator.MAX_STRIPES`? and also leads to the next point: 'adaptive'.

2. For the name of `AdaptivePoolingAllocator`, the magazines array can 'adaptively' expand in response to the threads contention, should we also make the magazines array be able to SHRINK when the threads contention becomes low? Then we should find a proper way to detect the threads contention level and choose a proper time-point to shrink the magazines array.

3. This PR comes as a 'by-product' of the previous points, which try to reduce threads collisions to improve the performance.

I'm very looking forwards to hear suggestions. @normanmaurer @chrisvest @franz1981

chrisvest commented:
> if we change the AdaptivePoolingAllocator.MAX_STRIPES value to 32, which equals with the JMH benchmark threads number, it will make a far better performance results than current PR: relatively 100% performance improvements.

This indicates the heuristic "twice the number of CPU cores" is too low? We also don't want to get too crazy, as systems have 192 cores or more.

> should we also make the magazines array be able to SHRINK when the threads contention becomes low?

The challenge is how to efficiently determine that contention is too low. It'd need to discern between right-sized and over-sized magazine array. We don't want to get into a mode where the magazines periodically grows and shrinks, back and forth.

laosijikaichele commented:
> This indicates the heuristic "twice the number of CPU cores" is too low?

The "twice the number of CPU cores" still correct.

But, it's a little different in the magazines array scenario. 
To make it simple, let's assume the computer has 4 cores, which means the `NettyRuntime.availableProcessors()` returns 4, then the `AdaptivePoolingAllocator.MAX_STRIPES` value is 8, assume we use 16 threads to run the allocating. As following graph shows:

<img width="865" alt="截屏2024-12-05 上午8 52 23" src="https://github.com/user-attachments/assets/c24e8d6b-9ef3-45fb-869c-05d9c87a914f">

As we can see, there are 4 cores running at the moment, but 'thread-id-8' and 'thread-id-16' get collision on magazine[0], and 'thread-id-1' and 'thread-id-9' get collision on magazine[1], which means there are only 2 cores are doing the real allocating work, which slowed down the performance, that's why when we change the `AdaptivePoolingAllocator.MAX_STRIPES` value to be equals with the JMH benchmark threads number, then the performance increased about 100%.
@chrisvest

laosijikaichele commented:
> Have you tried running the benchmark with the 'central queue' bit removed?

Some benchmark results:
1. Base: the current 4.1-branch code:
```
Benchmark                                                                         (size)   Mode  Cnt         Score         Error  Units
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00064  thrpt   20  44765852.405 ± 2158267.449  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00256  thrpt   20  50426422.617 ± 1625518.316  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   01024  thrpt   20  47962141.687 ± 1288288.181  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   04096  thrpt   20  47645849.622 ± 4059710.794  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00064  thrpt   20  45218005.401 ±  728359.822  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00256  thrpt   20  45705011.551 ± 1090326.331  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     01024  thrpt   20  40320395.856 ± 3147962.697  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     04096  thrpt   20  42041312.718 ± 1050791.102  ops/s
```
2. This PR:
```
Benchmark                                                                         (size)   Mode  Cnt         Score         Error  Units
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00064  thrpt   20  59228348.215 ± 1813789.557  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00256  thrpt   20  58336037.049 ± 1425381.453  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   01024  thrpt   20  55490531.505 ± 1725257.117  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   04096  thrpt   20  61100119.082 ± 2169315.148  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00064  thrpt   20  55866042.594 ± 2250531.728  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00256  thrpt   20  48192452.569 ± 2183073.224  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     01024  thrpt   20  52565337.762 ± 1783352.384  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     04096  thrpt   20  52622865.686 ± 3086720.804  ops/s
```
3. This PR: removed 'central queue' part:
```
Benchmark                                                                         (size)   Mode  Cnt         Score         Error  Units
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00064  thrpt   20  44214899.313 ± 1265912.293  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00256  thrpt   20  49909651.024 ± 2109215.889  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   01024  thrpt   20  42327685.343 ± 2913376.452  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   04096  thrpt   20  45835790.531 ± 2275833.994  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00064  thrpt   20  45252360.734 ± 3398428.018  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00256  thrpt   20  48799484.414 ± 3868804.633  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     01024  thrpt   20  46676173.107 ± 1816229.061  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     04096  thrpt   20  45385528.191 ± 1233630.633  ops/s
```
4. This PR: removed 'NEXT_IN_LINE' part:
```
Benchmark                                                                         (size)   Mode  Cnt         Score         Error  Units
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00064  thrpt   20  50789341.282 ± 2186182.068  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00256  thrpt   20  46420384.352 ± 1927074.344  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   01024  thrpt   20  49353791.249 ± 2145425.592  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   04096  thrpt   20  45599316.202 ± 1293919.331  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00064  thrpt   20  44419510.423 ± 1979362.192  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00256  thrpt   20  43488683.449 ± 2683059.025  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     01024  thrpt   20  45239700.847 ± 2573862.252  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     04096  thrpt   20  42303418.350 ± 1073820.743  ops/s
```
5. The current 4.1-branch code: Enlarged the `AdaptivePoolingAllocator.MAX_STRIPES` value to 32: 
```
Benchmark                                                                         (size)   Mode  Cnt         Score          Error  Units
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00064  thrpt   20  92044907.610 ±  5216575.193  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   00256  thrpt   20  86464537.298 ±  3298538.728  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   01024  thrpt   20  86211156.667 ±  1902848.102  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseDirectAdaptive   04096  thrpt   20  82572653.737 ±  2046694.213  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00064  thrpt   20  80209937.801 ± 13561768.134  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     00256  thrpt   20  89055010.211 ±  2713997.414  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     01024  thrpt   20  84096155.412 ±  1031039.806  ops/s
AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark.allocateReleaseHeapAdaptive     04096  thrpt   20  78620975.247 ±  2006552.766  ops/s
```

The benchmark-5 has the best performance, and the benchmark-2(current PR) has the second best performance.
Additionally, benchmark-5 has about 100% performance increased than the benchmark-1(base) benchmark.

laosijikaichele commented:
> using a local variable to register how many attempts are required before getting one 

I used to record these failed attempts counts and the proportion when I did the benchmark for the base branch a few days ago, the numbers seemed gradually decreased in a reasonable way, if I remembered it correctly. :)
Anyway I will re-run the benchmark and record those accurate numbers and share them later.

> why not trying to pick a random one to start with, after the failed attempt?

> Another approach can be to xorshift hash the thread id (to always hit the same), to better distribute it (which help with Virtual threads since their id tends to increase and not been reused!) - and than fallback with a random choice to re-start trying.

When the threads counts exceeded the `AdaptivePoolingAllocator.MAX_STRIPES`, there is not much we can do to avoid the threads get collisions on the same magazine, because we can not Interfere how the CPU schedule the threads, which illustrated in https://github.com/netty/netty/pull/14534#issuecomment-2518865160
@franz1981

laosijikaichele commented:
Currently, I can see two options to decrease the threads collisions:

1. Enlarge AdaptivePoolingAllocator.MAX_STRIPES to a proper value, which may need to do some evaluating/research to determine which value is proper, or we make it to be customizable by the users, then expand the magazines array when the contention level goes up(which currently 4.1-branch doing), and shrink the magazines array when contention level goes down, this may introduce some code implementation complexity.
2. Try to make some parallelism on the collision magazine, like the current PR did.

franz1981 commented:
I was suggesting something like this
```java
    private static int probe() {
        // Multiplicative Fibonacci hashing implementation
        // 0x9e3779b9 is the integral part of the Golden Ratio's fractional part 0.61803398875… (sqrt(5)-1)/2
        // multiplied by 2^32, which has the best possible scattering properties.
        int probe = (int) ((Thread.currentThread().getId() * 0x9e3779b9) & Integer.MAX_VALUE);
        // xorshift
        probe ^= probe << 13;
        probe ^= probe >>> 17;
        probe ^= probe << 5;
        return probe;
    }

    private AdaptiveByteBuf allocate(int size, int maxCapacity, Thread currentThread, AdaptiveByteBuf buf) {
        if (size <= MAX_CHUNK_SIZE) {
            int sizeBucket = AllocationStatistics.sizeBucket(size); // Compute outside of Magazine lock for better ILP.
            FastThreadLocal<Object> threadLocalMagazine = this.threadLocalMagazine;
            if (threadLocalMagazine != null && currentThread instanceof FastThreadLocalThread) {
                Object mag = threadLocalMagazine.get();
                if (mag != NO_MAGAZINE) {
                    Magazine magazine = (Magazine) mag;
                    if (buf == null) {
                        buf = magazine.newBuffer();
                    }
                    boolean allocated = magazine.tryAllocate(size, sizeBucket, maxCapacity, buf);
                    assert allocated : "Allocation of threadLocalMagazine must always succeed";
                    return buf;
                }
            }
            final int probe = probe();
            Magazine[] mags;
            int expansions = 0;
            do {
                mags = magazines;
                int mask = mags.length - 1;
                int magazineIndex = probe & mask;
                Magazine mag = mags[magazineIndex];
                if (buf == null) {
                    buf = mag.newBuffer();
                }
                if (mag.tryAllocate(size, sizeBucket, maxCapacity, buf)) {
                    return buf;
                }
                // If the first magazine failed, start from another random
                int randomIndex = (magazineIndex + ThreadLocalRandom.current().nextInt(1, mask + 1)) & mask;
                for (int i = 0; i < mask; i++) {
                    mag = mags[(randomIndex + i) & mask];
                    if (mag.tryAllocate(size, sizeBucket, maxCapacity, buf)) {
                        return buf;
                    }
                }
            } while (expansions++ <= EXPANSION_ATTEMPTS && tryExpandMagazines(mags.length));
        }
 ```

BEWARE: don't use it as it is, because I have probably messed up with the math after the first unsuccesful attempt (while `probe` should be decent already).

chrisvest commented:
I was pondering this PR and remembered something Franz told me recently, that hammering a concurrent data structure like this in a loop with no other work going on, will over-estimate the amount of contention we'll realistically see in practice.
For instance, you're not likely to see real code allocate and release buffers as aggressively as the benchmark used here.
So even though we can make changes that'll improve the benchmark, they might not translate into any tangible performance improvements in practice.

What I'm thinking is that, since the magazine stripes are the primary way we deal with contention, we should perhaps just introduce a system property for overriding the `MAX_STRIPES`. And maybe we want to expose a metric, or some mechanism, by which integrators can be informed if the allocator wants to expand the stripe array but the array is already at max size.

That would both be a smaller change, with a greater potential performance impact (that people can control), and opens a door to get more information about what contention is seen in practice.

franz1981 commented:
Can you try the PR I've sent by fixing a bit the logic of the

laosijikaichele commented:
> we should add BlackHole::consumeCPU between acquire/release and eventually handle burst/batches of acquire/release

@franz1981 
Did you mean the benchmark method should be like this?

   ```
 @Benchmark
    @Threads(32)
    public void allocateReleaseHeapAdaptive(Blackhole blackhole) {
        ByteBuf buf = adaptiveAllocator.heapBuffer(size);
        blackhole.consume(buf);
        blackhole.consume(buf.release());
    }
```

franz1981 commented:
Nope, you should have a int work parameter 

And call BlackHole.consumeCPU(work) if work > 0 between the buffer acquisition and release.

The work values could be 0, 10, 100.

Additionally you can have a new int batch field in the benchmark and a thread local state (in JMH) which expose a ArrayQueue<ByteBuf> field and you can acquire the buffers in batch (before the consumeCPU), place them in the thread local state stack/queue, and release them in burst draining from the stack in a sensible order (FIFO, LIFO? I think FIFO to be the way, if you think about the http 2 case...).
In this case the batch size can be 1, 10, 100

laosijikaichele commented:
@franz1981 
Forgive my stupidity, I am not quite understood it, may be you post the benchmark method code here? I will be happy to run it and record the results... :smile:

franz1981 commented:
No stupidity, producing code by text description is not trivial!
Start with this https://github.com/netty/netty/blob/2eb7fcef56ae82ab5ca5e1c9e03901bf94df204b/microbench/src/main/java/io/netty/microbench/buffer/SimpleByteBufPooledAllocatorBenchmark.java#L87

laosijikaichele commented:
> Blackhole.consumeCPU(tokens); 


Thanks @franz1981, I will try it, will update here once have progress.

franz1981 commented:
Additionally, I made this for JCTools (which I am a developer) - https://github.com/JCTools/JCTools/blob/a17b56f09cc9ec7d773707578c89f3ca9c11f968/jctools-benchmarks/src/main/java/org/jctools/jmh/latency/QueueAsPoolBurstCost.java#L91

Which add the batching I have mentioned.

I suggest to do the same here, because it's a type of pattern we have in Netty: sadly there is still a missing bit which I would like to capture, but let's talk about it when you modify the benchmark with both these changes.

laosijikaichele commented:
> And call BlackHole.consumeCPU(work) (a static method!) if work > 0 between the buffer acquisition and release.
> The work values could be 0, 10, 100.

Done this part so far.

1. The 4.1-branch results:


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/jason/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/jason/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


4.1-branch | size = 123, tokens = 0 | size = 123, tokens = 5 | size = 123, tokens = 10 | size = 123, tokens = 100 | size = 1234, tokens = 0 | size = 1234, tokens = 5 | size = 1234, tokens = 10 | size = 1234, tokens = 100 | size = 12345, tokens = 0 | size = 12345, tokens = 5 | size = 12345, tokens = 10 | size = 12345, tokens = 100 | size = 123456, tokens = 0 | size = 123456, tokens = 5 | size = 123456, tokens = 10 | size = 123456, tokens = 100 | size = 1234567, tokens = 0 | size = 1234567, tokens = 5 | size = 1234567, tokens = 10 | size = 1234567, tokens = 100
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
total_allocate_counts | 214254916 | 217367769 | 210315954 | 235856697 | 216316100 | 213340946 | 204965371 | 201097085 | 202935425 | 212224972 | 204924222 | 192961743 | 7904007 | 8069286 | 8004592 | 8024263 | 245131381 | 246621681 | 210477630 | 182198975
success_counts_at_expansions_0_pointerBump_0 | 184080241 | 189316486 | 183654849 | 203136810 | 187533439 | 185616076 | 175082056 | 175519979 | 174208401 | 187253041 | 175014753 | 165460919 | 1952710 | 1794491 | 1869813 | 2039150 | 174484268 | 184802619 | 142790563 | 137149366
success_counts_at_expansions_0_pointerBump_1 | 20931962 | 19772935 | 19157378 | 23800493 | 22061678 | 19644826 | 21439797 | 19167610 | 21338545 | 19120088 | 22638007 | 21700909 | 1350511 | 1530030 | 1483877 | 1448596 | 42254729 | 39386349 | 45216521 | 30139261
success_counts_at_expansions_0_pointerBump_2 | 5883271 | 5688887 | 5307883 | 5847105 | 4659207 | 5520573 | 5262642 | 4098660 | 4933545 | 4366556 | 4969204 | 4216007 | 1077966 | 1153831 | 1075405 | 1110443 | 20040673 | 14059802 | 16064265 | 10259393
success_counts_at_expansions_0_pointerBump_3 | 1694417 | 1547865 | 1294298 | 1631441 | 1298755 | 1449541 | 1984094 | 1269280 | 1592985 | 1045316 | 1385642 | 1064650 | 800015 | 874659 | 839396 | 817692 | 7542723 | 7676952 | 5717658 | 4119370
success_counts_at_expansions_1_pointerBump_0 | 75652 | 54774 | 62829 | 74854 | 72923 | 63926 | 94666 | 53183 | 70612 | 54225 | 80445 | 50907 | 2685 | 2870 | 2987 | 6373 | 188151 | 160636 | 144788 | 100279
success_counts_at_expansions_1_pointerBump_1 | 50441 | 37289 | 40988 | 44135 | 45676 | 45385 | 60907 | 34073 | 52868 | 37769 | 55073 | 37079 | 1539 | 2055 | 1736 | 3965 | 133003 | 112661 | 114382 | 70958
success_counts_at_expansions_1_pointerBump_2 | 35660 | 33191 | 31716 | 33811 | 26827 | 30763 | 43584 | 20207 | 30560 | 23425 | 35438 | 20188 | 1054 | 1357 | 1212 | 2503 | 80553 | 61361 | 67305 | 38189
success_counts_at_expansions_1_pointerBump_3 | 17489 | 16216 | 13875 | 15907 | 14286 | 14637 | 26270 | 13500 | 17302 | 10660 | 18933 | 11230 | 701 | 803 | 711 | 1577 | 28182 | 21509 | 23770 | 14418
success_counts_at_expansions_2_pointerBump_0 | 14320 | 8268 | 9330 | 11088 | 11406 | 10858 | 13971 | 9239 | 9820 | 7269 | 10278 | 6453 | 972 | 972 | 1123 | 1811 | 22855 | 18272 | 17488 | 13912
success_counts_at_expansions_2_pointerBump_1 | 11628 | 7773 | 6086 | 8084 | 8312 | 8336 | 10264 | 6281 | 9173 | 5058 | 8836 | 5415 | 639 | 849 | 863 | 1375 | 17469 | 13934 | 15396 | 10000
success_counts_at_expansions_2_pointerBump_2 | 9994 | 8808 | 7264 | 7866 | 5857 | 6718 | 11113 | 5217 | 7793 | 4904 | 7553 | 4378 | 653 | 721 | 753 | 1265 | 14039 | 10807 | 12184 | 7357
success_counts_at_expansions_2_pointerBump_3 | 11166 | 6845 | 4810 | 9289 | 5516 | 6125 | 10617 | 5609 | 7896 | 3762 | 7058 | 4293 | 572 | 648 | 696 | 1013 | 8381 | 7064 | 7320 | 6040
success_counts_at_expansions_3_pointerBump_0 | 11627 | 6609 | 6513 | 8946 | 9001 | 8115 | 10053 | 6682 | 7068 | 5903 | 6427 | 3922 | 859 | 883 | 1070 | 1736 | 11649 | 10865 | 10187 | 9458
success_counts_at_expansions_3_pointerBump_1 | 9610 | 5086 | 3827 | 6453 | 5788 | 5965 | 6811 | 5463 | 7193 | 3423 | 5484 | 3495 | 624 | 798 | 872 | 1280 | 9741 | 8522 | 9527 | 7048
success_counts_at_expansions_3_pointerBump_2 | 8965 | 8344 | 6380 | 5981 | 4662 | 5662 | 9095 | 4627 | 6397 | 3311 | 5723 | 2956 | 638 | 720 | 767 | 1160 | 8790 | 8139 | 8603 | 5994
success_counts_at_expansions_3_pointerBump_3 | 8918 | 8240 | 5593 | 8862 | 5043 | 6608 | 9806 | 6170 | 8247 | 3956 | 6114 | 4123 | 656 | 737 | 743 | 1045 | 6900 | 6776 | 6898 | 6182
success_counts_at_allocateFallback | 1399555 | 840153 | 702335 | 1205572 | 547724 | 896832 | 889625 | 871305 | 627020 | 276306 | 669254 | 364819 | 2711213 | 2702862 | 2722568 | 2583279 | 279275 | 255413 | 250775 | 241750



</body>

</html>

2. Current PR results:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/jason/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/jason/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


current-PR | size = 123, tokens = 0 | size = 123, tokens = 5 | size = 123, tokens = 10 | size = 123, tokens = 100 | size = 1234, tokens = 0 | size = 1234, tokens = 5 | size = 1234, tokens = 10 | size = 1234, tokens = 100 | size = 12345, tokens = 0 | size = 12345, tokens = 5 | size = 12345, tokens = 10 | size = 12345, tokens = 100 | size = 123456, tokens = 0 | size = 123456, tokens = 5 | size = 123456, tokens = 10 | size = 123456, tokens = 100 | size = 1234567, tokens = 0 | size = 1234567, tokens = 5 | size = 1234567, tokens = 10 | size = 1234567, tokens = 100
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
total_allocate_counts | 301434249 | 200717576 | 207727540 | 202524927 | 213958453 | 311556773 | 210252727 | 191639942 | 200670408 | 211401452 | 297834767 | 233786702 | 321846873 | 330986127 | 310886936 | 297527400 | 217752753 | 203583541 | 305363823 | 181861413
success_counts_at_expansions_0_pointerBump_0 | 301430185 | 200712828 | 207719733 | 202522574 | 213956583 | 311554613 | 210251380 | 191637844 | 200667392 | 211398109 | 297830477 | 233779925 | 321846641 | 330985868 | 310886623 | 297527105 | 217458203 | 203355926 | 305162621 | 181702829
success_counts_at_expansions_0_pointerBump_1 | 3124 | 2775 | 5196 | 1290 | 1432 | 1850 | 1147 | 1550 | 2941 | 3220 | 4049 | 5936 | 88 | 90 | 95 | 66 | 157332 | 163216 | 124054 | 103550
success_counts_at_expansions_0_pointerBump_2 | 573 | 1129 | 852 | 630 | 263 | 181 | 102 | 420 | 56 | 97 | 215 | 822 | 40 | 17 | 44 | 44 | 81326 | 45304 | 47990 | 40038
success_counts_at_expansions_0_pointerBump_3 | 133 | 347 | 824 | 395 | 96 | 60 | 64 | 57 | 17 | 11 | 3 | 7 | 25 | 39 | 28 | 12 | 51291 | 14333 | 20884 | 10902
success_counts_at_expansions_1_pointerBump_0 | 43 | 57 | 241 | 8 | 17 | 35 | 2 | 10 | 0 | 4 | 20 | 2 | 1 | 6 | 13 | 11 | 891 | 1431 | 3396 | 803
success_counts_at_expansions_1_pointerBump_1 | 28 | 48 | 105 | 2 | 8 | 1 | 0 | 3 | 0 | 1 | 0 | 0 | 1 | 8 | 6 | 3 | 709 | 730 | 750 | 423
success_counts_at_expansions_1_pointerBump_2 | 29 | 123 | 30 | 5 | 7 | 4 | 1 | 2 | 0 | 0 | 0 | 1 | 1 | 4 | 4 | 5 | 226 | 364 | 527 | 334
success_counts_at_expansions_1_pointerBump_3 | 15 | 29 | 183 | 11 | 5 | 4 | 14 | 5 | 0 | 0 | 0 | 0 | 2 | 7 | 1 | 2 | 470 | 148 | 1222 | 335
success_counts_at_expansions_2_pointerBump_0 | 15 | 16 | 16 | 3 | 3 | 1 | 0 | 3 | 0 | 0 | 0 | 0 | 6 | 2 | 8 | 4 | 111 | 159 | 209 | 104
success_counts_at_expansions_2_pointerBump_1 | 6 | 12 | 40 | 1 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 2 | 1 | 7 | 1 | 2 | 336 | 135 | 133 | 107
success_counts_at_expansions_2_pointerBump_2 | 9 | 30 | 13 | 0 | 15 | 0 | 0 | 1 | 0 | 0 | 0 | 0 | 1 | 2 | 2 | 0 | 43 | 51 | 251 | 28
success_counts_at_expansions_2_pointerBump_3 | 5 | 15 | 53 | 5 | 1 | 1 | 7 | 1 | 0 | 1 | 0 | 0 | 0 | 6 | 1 | 1 | 63 | 48 | 108 | 143
success_counts_at_expansions_3_pointerBump_0 | 11 | 9 | 9 | 0 | 1 | 2 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 2 | 2 | 2 | 60 | 55 | 57 | 61
success_counts_at_expansions_3_pointerBump_1 | 9 | 2 | 27 | 0 | 2 | 0 | 0 | 1 | 0 | 0 | 0 | 1 | 1 | 3 | 2 | 0 | 110 | 63 | 27 | 81
success_counts_at_expansions_3_pointerBump_2 | 3 | 2 | 19 | 0 | 2 | 1 | 1 | 2 | 0 | 0 | 0 | 0 | 0 | 1 | 0 | 2 | 23 | 14 | 75 | 25
success_counts_at_expansions_3_pointerBump_3 | 4 | 1 | 17 | 1 | 2 | 2 | 3 | 0 | 0 | 0 | 0 | 0 | 1 | 2 | 1 | 1 | 37 | 15 | 89 | 80
success_counts_at_allocateFallback | 57 | 153 | 182 | 2 | 15 | 17 | 6 | 41 | 2 | 7 | 3 | 6 | 63 | 63 | 105 | 140 | 1522 | 1549 | 1430 | 1570



</body>

</html>


The benchmark method code looks like:

```
private static final AdaptiveByteBufAllocator adaptiveAllocator =
            new AdaptiveByteBufAllocator(true, false);

    @Param({
            "123",
            "1234",
            "12345",
            "123456",
            "1234567",
    })
    public int size;

    @Param({
            "0",
            "5",
            "10",
            "100",
    })
    public long tokens;

    public AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark2() {
        super(false, true);
    }

    @Benchmark
    @Threads(32)
    public void allocateReleaseHeapAdaptive(Blackhole blackhole) {
        ByteBuf buf = adaptiveAllocator.heapBuffer(size);
        if (tokens > 0) {
            Blackhole.consumeCPU(tokens);
        }
        blackhole.consume(buf.release());
    }
```
@franz1981

franz1981 commented:
As you can see with tokens at 100 the before/after probably won't show much difference (and that's what real world will look like). Do you have the ns/op?

Ok, about using the snippet I suggested at https://github.com/netty/netty/pull/14534#issuecomment-2520030722 (to be modified)?

laosijikaichele commented:
> about using the snippet I suggested at https://github.com/netty/netty/pull/14534#issuecomment-2520030722 (to be modified)?

The results, with 4.1-branch + 'randomProb':
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/jason/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/jason/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


4.1-branch, with   randomProb | size = 1234, tokens = 0 | size = 1234, tokens = 100
-- | -- | --
total_allocate_counts | 219591307 | 181831112
expansions_0_before_pointerBump | 155150620 | 142771233
expansions_1_before_pointerBump | 6931 | 1009
expansions_2_before_pointerBump | 1244 | 197
expansions_3_before_pointerBump | 1244 | 197
success_counts_at_expansions_0_pointerBump_0 | 42404185 | 29689043
success_counts_at_expansions_0_pointerBump_1 | 12606060 | 6464614
success_counts_at_expansions_0_pointerBump_2 | 4960809 | 1934550
success_counts_at_expansions_0_pointerBump_3 | 2114302 | 600568
success_counts_at_expansions_1_pointerBump_0 | 7241 | 597
success_counts_at_expansions_1_pointerBump_1 | 6068 | 507
success_counts_at_expansions_1_pointerBump_2 | 5106 | 429
success_counts_at_expansions_1_pointerBump_3 | 4814 | 472
success_counts_at_expansions_2_pointerBump_0 | 1337 | 163
success_counts_at_expansions_2_pointerBump_1 | 1262 | 157
success_counts_at_expansions_2_pointerBump_2 | 1173 | 142
success_counts_at_expansions_2_pointerBump_3 | 1176 | 148
success_counts_at_expansions_3_pointerBump_0 | 543 | 99
success_counts_at_expansions_3_pointerBump_1 | 539 | 67
success_counts_at_expansions_3_pointerBump_2 | 516 | 82
success_counts_at_expansions_3_pointerBump_3 | 510 | 87
success_counts_at_allocateFallback | 15520 | 6007



</body>

</html>

Seems there is no performance improvements than the 4.1-branch code, the first successful alloc distribution(expansions_0_before_pointerBump) is about 70%~78%. 

@franz1981 
 
Honestly, I have tried the similar random method to observe the performance change few days ago, it showed the similar results with this 'randomProb' implementation.

Again, as I have said, there is not much we can do when the threads counts exceeded `MAX_STRIPES`, which illustrated in 
https://github.com/netty/netty/pull/14534#issuecomment-2518865160.

franz1981 commented:
Thanks to have tried, although it is a bit weird, since I would have expected a fairest distribution...

Said that..

> As you can see with tokens at 100 the before/after probably won't show much difference (and that's what real world will look like)

Which means that, as @chrisvest suggested; this is an improvement which disappear as the user does something (token 100), if I read correctly your data; if not, can you share the ns/op before/after your changes?

laosijikaichele commented:
> expansions_0-1-2-3_before_pointerBump

These correspond to your code snippet in https://github.com/netty/netty/pull/14534#issuecomment-2520030722 which introduced an extra allocation operation before the `for-loop`:
```
                if (mag.tryAllocate(size, sizeBucket, maxCapacity, buf)) {
                    // Here: expansions_0-1-2-3_before_pointerBump.
                    return buf;
                }
                int randomIndex = (magazineIndex + ThreadLocalRandom.current().nextInt(1, mask + 1)) & mask;
                for (int i = 0; i < mask; i++) {
                    mag = mags[(randomIndex + i) & mask];
                    if (mag.tryAllocate(size, sizeBucket, maxCapacity, buf)) {
                        // Here: success_counts_at_expansions_0_pointerBump_0-1-2-3....-mask
                        return buf;
                    }
                }
```
Additionally, I do not record all of the pointerBump like [0,1,2,3 - mask] but only expansions_pointerBump_0-1-2-3.

laosijikaichele commented:
> this is an improvement which disappear as the user does something (token 100)

Sorry I didn't get the meaning, could you please explain it more? @franz1981

laosijikaichele commented:
> As said, the distribution under contention, can be very relevant, but as @chrisvest said, we should add BlackHole::consumeCPU between acquire/release and eventually handle burst/batches of acquire/release

To do this, I modified the benchmark code to be [AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark2](https://github.com/laosijikaichele/netty/blob/tmp/microbench/src/main/java/io/netty/buffer/AdaptiveByteBufAllocatorConcurrentNoCacheBenchmark2.java)

The following are the benchmark results:
1. 4.1-branch:
```
Benchmark                      (size)  (tokens)  Mode  Cnt         Score         Error  Units
allocateReleaseHeapAdaptive      123         0  thrpt   20  49600569.565 ± 3541877.971  ops/s
allocateReleaseHeapAdaptive      123         5  thrpt   20  52596593.636 ± 1472694.081  ops/s
allocateReleaseHeapAdaptive      123        10  thrpt   20  43768030.709 ± 1184420.657  ops/s
allocateReleaseHeapAdaptive      123       100  thrpt   20  18626691.915 ±  476974.252  ops/s
allocateReleaseHeapAdaptive     1234         0  thrpt   20  55033762.901 ± 1610645.207  ops/s
allocateReleaseHeapAdaptive     1234         5  thrpt   20  49146623.483 ± 1101063.739  ops/s
allocateReleaseHeapAdaptive     1234        10  thrpt   20  48111169.236 ± 1509283.542  ops/s
allocateReleaseHeapAdaptive     1234       100  thrpt   20  18762242.666 ±  333892.841  ops/s
allocateReleaseHeapAdaptive    12345         0  thrpt   20  52435911.748 ± 1850134.115  ops/s
allocateReleaseHeapAdaptive    12345         5  thrpt   20  49430211.528 ± 2192686.978  ops/s
allocateReleaseHeapAdaptive    12345        10  thrpt   20  42536581.878 ± 3363863.074  ops/s
allocateReleaseHeapAdaptive    12345       100  thrpt   20  18799263.626 ±  744005.641  ops/s
allocateReleaseHeapAdaptive   123456         0  thrpt   20    365191.379 ±    4097.401  ops/s
allocateReleaseHeapAdaptive   123456         5  thrpt   20    366485.563 ±    5019.905  ops/s
allocateReleaseHeapAdaptive   123456        10  thrpt   20    358996.492 ±    8624.480  ops/s
allocateReleaseHeapAdaptive   123456       100  thrpt   20    367667.187 ±    5783.807  ops/s
allocateReleaseHeapAdaptive  1234567         0  thrpt   20  37227841.126 ± 1392361.728  ops/s
allocateReleaseHeapAdaptive  1234567         5  thrpt   20  35079905.508 ± 1633816.587  ops/s
allocateReleaseHeapAdaptive  1234567        10  thrpt   20  32707449.358 ± 2821133.525  ops/s
allocateReleaseHeapAdaptive  1234567       100  thrpt   20  13373001.783 ±  648725.973  ops/s
```
2. 4.1-branch with @franz1981 'randomProb' implementation:
```
Benchmark                     (size)  (tokens)   Mode  Cnt         Score         Error  Units
allocateReleaseHeapAdaptive      123         0  thrpt   20  32012619.071 ± 2749938.105  ops/s
allocateReleaseHeapAdaptive      123         5  thrpt   20  31282002.783 ± 1569598.159  ops/s
allocateReleaseHeapAdaptive      123        10  thrpt   20  30366831.990 ± 1584817.658  ops/s
allocateReleaseHeapAdaptive      123       100  thrpt   20  15742239.592 ±  346598.796  ops/s
allocateReleaseHeapAdaptive     1234         0  thrpt   20  34656269.980 ± 1711087.653  ops/s
allocateReleaseHeapAdaptive     1234         5  thrpt   20  31310085.694 ± 1533949.376  ops/s
allocateReleaseHeapAdaptive     1234        10  thrpt   20  29051633.496 ±  918231.946  ops/s
allocateReleaseHeapAdaptive     1234       100  thrpt   20  16231425.631 ±  332671.316  ops/s
allocateReleaseHeapAdaptive    12345         0  thrpt   20  31237676.285 ± 1559837.123  ops/s
allocateReleaseHeapAdaptive    12345         5  thrpt   20  27743366.146 ± 1013648.387  ops/s
allocateReleaseHeapAdaptive    12345        10  thrpt   20  28474031.904 ± 1601307.744  ops/s
allocateReleaseHeapAdaptive    12345       100  thrpt   20  15359808.346 ±  471578.403  ops/s
allocateReleaseHeapAdaptive   123456         0  thrpt   20    364882.114 ±    7803.780  ops/s
allocateReleaseHeapAdaptive   123456         5  thrpt   20    363344.221 ±    5323.685  ops/s
allocateReleaseHeapAdaptive   123456        10  thrpt   20    361630.220 ±    6682.916  ops/s
allocateReleaseHeapAdaptive   123456       100  thrpt   20    369328.294 ±    4427.815  ops/s
allocateReleaseHeapAdaptive  1234567         0  thrpt   20  20225717.622 ± 2757264.861  ops/s
allocateReleaseHeapAdaptive  1234567         5  thrpt   20  17191142.957 ± 1638853.157  ops/s
allocateReleaseHeapAdaptive  1234567        10  thrpt   20  15318177.193 ± 1202633.061  ops/s
allocateReleaseHeapAdaptive  1234567       100  thrpt   20  10775905.274 ±  343227.446  ops/s
```
3. Current PR:
```
Benchmark                       (size)  (tokens)  Mode  Cnt          Score          Error  Units
allocateReleaseHeapAdaptive      123         0  thrpt   20   76775268.663 ±  1712651.792  ops/s
allocateReleaseHeapAdaptive      123         5  thrpt   20   65730696.539 ±  1664060.591  ops/s
allocateReleaseHeapAdaptive      123        10  thrpt   20   58170042.646 ±  6412294.126  ops/s
allocateReleaseHeapAdaptive      123       100  thrpt   20   20419676.377 ±   242858.555  ops/s
allocateReleaseHeapAdaptive     1234         0  thrpt   20   61923322.493 ±  6520030.367  ops/s
allocateReleaseHeapAdaptive     1234         5  thrpt   20   60468843.566 ±  3128129.438  ops/s
allocateReleaseHeapAdaptive     1234        10  thrpt   20   58892722.014 ±  2503542.297  ops/s
allocateReleaseHeapAdaptive     1234       100  thrpt   20   19984648.170 ±   410064.453  ops/s
allocateReleaseHeapAdaptive    12345         0  thrpt   20   65496279.721 ±  5480641.901  ops/s
allocateReleaseHeapAdaptive    12345         5  thrpt   20   57451786.964 ±  2827528.935  ops/s
allocateReleaseHeapAdaptive    12345        10  thrpt   20   52137740.145 ±  2594627.374  ops/s
allocateReleaseHeapAdaptive    12345       100  thrpt   20   19507948.566 ±   276660.639  ops/s
allocateReleaseHeapAdaptive   123456         0  thrpt   20  123998109.885 ± 13292543.621  ops/s
allocateReleaseHeapAdaptive   123456         5  thrpt   20  101797528.387 ±  8081467.032  ops/s
allocateReleaseHeapAdaptive   123456        10  thrpt   20  102573508.117 ±  8332524.166  ops/s
allocateReleaseHeapAdaptive   123456       100  thrpt   20   25848586.858 ±  1204460.494  ops/s
allocateReleaseHeapAdaptive  1234567         0  thrpt   20   55967554.848 ±  1006042.586  ops/s
allocateReleaseHeapAdaptive  1234567         5  thrpt   20   55495435.614 ±  4690964.986  ops/s
allocateReleaseHeapAdaptive  1234567        10  thrpt   20   48915941.704 ±  2846835.362  ops/s
allocateReleaseHeapAdaptive  1234567       100  thrpt   20   18897445.489 ±   435000.581  ops/s
```
We can see that current PR has the best performance, and 4.1-branch is the second.

It's interesting that when the size goes to '123456', the current-PR make a **HUGE** performance improvement than 4.1-branch, around **70 - 250 TIMES** performance improvement. @franz1981 @chrisvest

laosijikaichele commented:
> Which means that, as @chrisvest suggested; this is an improvement which disappear as the user does something (token 100), if I read correctly your data.

I think you have mis-read the benchmark results, based on all the benchmark results so far, the improvement of this PR has never been disappeared. :smiley: @franz1981

franz1981 commented:
Yep @laosijikaichele but I want to try it in some box and better understand what's going on first; that's why I asked for some clarification about the counters and metrics and their meaning

laosijikaichele commented:
But, there is a point still need to be added:
This PR only kicks in when threads collide on the same magazine, so if there is no threads collision on the same magazine, we can say this PR does not take effect on the performance improvement.

franz1981 commented:
Did you modified the code to make it happen? 
If not I can think about it

laosijikaichele commented:
> Did you modified the code to make it happen? If not I can think about it

Did you mean 'This PR only kicks in when threads collide on the same magazine'? 
This PR already done this.
@franz1981

normanmaurer commented:
Yeah please

laosijikaichele commented:
Sorry, wrongly closed it when trying to resolve the conflicts, I will re-open it or file a new PR soon.

laosijikaichele commented:
New PR was filed, please move to https://github.com/netty/netty/pull/14594, sorry for the disturbing.

laosijikaichele commented:
> Can we check the numbers after the many latest changes? @normanmaurer wdyt?

Will re-do the test...

laosijikaichele commented:
> > Can we check the numbers after the many latest changes? @normanmaurer wdyt?
> 
> Will re-do the test...

JMH results updated: https://github.com/netty/netty/pull/14594#issuecomment-2550288966
